File: packages/sdk/src/economic/budget-engine.ts
/**
 * Local budget engine — in-memory, session scope only.
 *
 * Implements the BudgetEngine interface for standalone SDK mode.
 * Only materializes 'session' scope; agent/user/global require
 * CloudBudgetEngine (platform).
 *
 * @module economic/budget-engine
 */

import type {
  BudgetEngine,
  BudgetCheckResult,
  BudgetScope,
  EconomicBudgetConfig,
  EconomicBudgetStatus,
  EconomicDenialDetails,
} from './types.js';
import type { Logger } from '../utils/logger.js';

interface BudgetState {
  spent: number;
  limit: number;
  currency: string;
  approval_threshold?: number;
}

export interface LocalBudgetEngineOptions {
  budgets: EconomicBudgetConfig[];
  logger: Logger;
}

export class LocalBudgetEngine implements BudgetEngine {
  private readonly state: Map<BudgetScope, BudgetState> = new Map();
  private readonly logger: Logger;

  constructor(options: LocalBudgetEngineOptions) {
    this.logger = options.logger;

    for (const budget of options.budgets) {
      if (budget.window !== 'session') {
        this.logger.warn(
          'LocalBudgetEngine only supports session scope — ' +
          `ignoring ${budget.scope} budget with window '${budget.window}'. ` +
          'Use Veto Cloud for agent/user/global scopes.',
          { scope: budget.scope, window: budget.window }
        );
        continue;
      }

      this.state.set(budget.scope, {
        spent: 0,
        limit: budget.limit,
        currency: budget.currency,
        approval_threshold: budget.approval_threshold,
      });
    }
  }

  check(cost: number, currency: string, scope: BudgetScope): BudgetCheckResult {
    const safeCost = this.sanitizeCost(cost);
    const budget = this.state.get(scope);

    if (!budget) {
      // No budget configured for this scope — allow by default
      return { allowed: true, decision: 'allow' };
    }

    // Currency mismatch check
    if (currency.toUpperCase() !== budget.currency.toUpperCase()) {
      return {
        allowed: false,
        decision: 'deny',
        denial: this.buildDenial('currency_mismatch', safeCost, currency, budget, scope),
      };
    }

    const remaining = budget.limit - budget.spent;

    // Budget exceeded
    if (safeCost > remaining) {
      this.logger.warn('Budget would be exceeded', {
        scope, cost: safeCost, spent: budget.spent, limit: budget.limit,
      });
      return {
        allowed: false,
        decision: 'deny',
        denial: this.buildDenial('budget_exceeded', safeCost, currency, budget, scope),
      };
    }

    // Approval threshold
    if (budget.approval_threshold !== undefined && safeCost > budget.approval_threshold) {
      this.logger.info('Cost exceeds approval threshold', {
        scope, cost: safeCost, threshold: budget.approval_threshold,
      });
      return {
        allowed: false,
        decision: 'require_approval',
        denial: this.buildDenial('approval_required', safeCost, currency, budget, scope),
      };
    }

    return { allowed: true, decision: 'allow' };
  }

  reserve(cost: number, currency: string, scope: BudgetScope): BudgetCheckResult {
    const result = this.check(cost, currency, scope);
    if (!result.allowed) return result;

    const safeCost = this.sanitizeCost(cost);
    const budget = this.state.get(scope);
    if (budget && safeCost > 0) {
      budget.spent += safeCost;
      this.logger.debug('Budget reserved', {
        scope, cost: safeCost, totalSpent: budget.spent,
        remaining: budget.limit - budget.spent,
      });
    }

    return result;
  }

  record(cost: number, _currency: string, scope: BudgetScope): void {
    const safeCost = this.sanitizeCost(cost);
    const budget = this.state.get(scope);
    if (budget && safeCost > 0) {
      budget.spent += safeCost;
      this.logger.debug('Budget charge recorded', {
        scope, cost: safeCost, totalSpent: budget.spent,
        remaining: budget.limit - budget.spent,
      });
    }
  }

  refund(amount: number, scope: BudgetScope): void {
    if (amount <= 0) return;
    const budget = this.state.get(scope);
    if (budget) {
      budget.spent = Math.max(0, budget.spent - amount);
      this.logger.debug('Budget refunded', {
        scope, amount, totalSpent: budget.spent,
        remaining: budget.limit - budget.spent,
      });
    }
  }

  getStatus(scope: BudgetScope): EconomicBudgetStatus | null {
    const budget = this.state.get(scope);
    if (!budget) return null;
    return {
      scope,
      spent: budget.spent,
      limit: budget.limit,
      remaining: Math.max(0, budget.limit - budget.spent),
      currency: budget.currency,
    };
  }

  reset(scope: BudgetScope): void {
    const budget = this.state.get(scope);
    if (budget) {
      const previous = budget.spent;
      budget.spent = 0;
      this.logger.debug('Budget reset', { scope, previousSpent: previous });
    }
  }

  private sanitizeCost(cost: number): number {
    if (!Number.isFinite(cost) || cost < 0) return 0;
    return cost;
  }

  private buildDenial(
    reason: EconomicDenialDetails['reason'],
    cost: number,
    currency: string,
    budget: BudgetState,
    scope: BudgetScope,
  ): EconomicDenialDetails {
    return {
      reason,
      cost,
      currency,
      budget_scope: scope,
      budget_limit: budget.limit,
      budget_spent: budget.spent,
      budget_remaining: Math.max(0, budget.limit - budget.spent),
      approval_threshold: budget.approval_threshold,
    };
  }
}

File: packages/sdk-python/veto/economic/budget_engine.py
"""Local in-memory budget engine for economic authorization."""

from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Optional

from veto.economic.types import (
    BudgetCheckResult,
    BudgetScope,
    EconomicBudgetConfig,
    EconomicBudgetStatus,
    EconomicDenialDetails,
    EconomicDenialReason,
)
from veto.utils.logger import Logger


@dataclass
class _BudgetState:
    spent: float
    limit: float
    currency: str
    approval_threshold: Optional[float] = None


class LocalBudgetEngine:
    """Session-only in-memory budget engine.

    TypeScript intentionally keeps local SDK budgets scoped to ``session``.
    Agent/user/global windows need the platform because they need shared state.
    """

    def __init__(
        self,
        *,
        budgets: list[EconomicBudgetConfig],
        logger: Logger,
    ) -> None:
        self._logger = logger
        self._state: dict[BudgetScope, _BudgetState] = {}

        for budget in budgets:
            if budget.scope != "session":
                self._logger.warn(
                    "LocalBudgetEngine only supports session scope — "
                    f"ignoring {budget.scope} budget. Use Veto Cloud for "
                    "agent/user/global scopes.",
                    {"scope": budget.scope, "window": budget.window},
                )
                continue

            self._state[budget.scope] = _BudgetState(
                spent=0,
                limit=budget.limit,
                currency=budget.currency,
                approval_threshold=budget.approval_threshold,
            )

    def check(
        self,
        cost: float,
        currency: str,
        scope: BudgetScope,
    ) -> BudgetCheckResult:
        safe_cost = self._sanitize_cost(cost)
        budget = self._state.get(scope)
        if budget is None:
            return BudgetCheckResult(allowed=True, decision="allow")

        if currency.upper() != budget.currency.upper():
            return BudgetCheckResult(
                allowed=False,
                decision="deny",
                denial=self._build_denial(
                    "currency_mismatch", safe_cost, currency, budget, scope
                ),
            )

        remaining = budget.limit - budget.spent
        if safe_cost > remaining:
            self._logger.warn(
                "Budget would be exceeded",
                {
                    "scope": scope,
                    "cost": safe_cost,
                    "spent": budget.spent,
                    "limit": budget.limit,
                },
            )
            return BudgetCheckResult(
                allowed=False,
                decision="deny",
                denial=self._build_denial(
                    "budget_exceeded", safe_cost, currency, budget, scope
                ),
            )

        if (
            budget.approval_threshold is not None
            and safe_cost > budget.approval_threshold
        ):
            self._logger.info(
                "Cost exceeds approval threshold",
                {
                    "scope": scope,
                    "cost": safe_cost,
                    "threshold": budget.approval_threshold,
                },
            )
            return BudgetCheckResult(
                allowed=False,
                decision="require_approval",
                denial=self._build_denial(
                    "approval_required", safe_cost, currency, budget, scope
                ),
            )

        return BudgetCheckResult(allowed=True, decision="allow")

    def reserve(
        self,
        cost: float,
        currency: str,
        scope: BudgetScope,
    ) -> BudgetCheckResult:
        result = self.check(cost, currency, scope)
        if not result.allowed:
            return result

        safe_cost = self._sanitize_cost(cost)
        budget = self._state.get(scope)
        if budget is not None and safe_cost > 0:
            budget.spent += safe_cost
            self._logger.debug(
                "Budget reserved",
                {
                    "scope": scope,
                    "cost": safe_cost,
                    "totalSpent": budget.spent,
                    "remaining": budget.limit - budget.spent,
                },
            )

        return result

    def record(self, cost: float, _currency: str, scope: BudgetScope) -> None:
        safe_cost = self._sanitize_cost(cost)
        budget = self._state.get(scope)
        if budget is not None and safe_cost > 0:
            budget.spent += safe_cost
            self._logger.debug(
                "Budget charge recorded",
                {
                    "scope": scope,
                    "cost": safe_cost,
                    "totalSpent": budget.spent,
                    "remaining": budget.limit - budget.spent,
                },
            )

    def refund(self, amount: float, scope: BudgetScope) -> None:
        if amount <= 0:
            return

        budget = self._state.get(scope)
        if budget is None:
            return

        budget.spent = max(0, budget.spent - amount)
        self._logger.debug(
            "Budget refunded",
            {
                "scope": scope,
                "amount": amount,
                "totalSpent": budget.spent,
                "remaining": budget.limit - budget.spent,
            },
        )

    def get_status(self, scope: BudgetScope) -> Optional[EconomicBudgetStatus]:
        budget = self._state.get(scope)
        if budget is None:
            return None
        return EconomicBudgetStatus(
            scope=scope,
            spent=budget.spent,
            limit=budget.limit,
            remaining=max(0, budget.limit - budget.spent),
            currency=budget.currency,
        )

    # TypeScript-style alias for callers translating examples directly.
    def getStatus(self, scope: BudgetScope) -> Optional[EconomicBudgetStatus]:
        return self.get_status(scope)

    def reset(self, scope: BudgetScope) -> None:
        budget = self._state.get(scope)
        if budget is None:
            return

        previous = budget.spent
        budget.spent = 0
        self._logger.debug("Budget reset", {"scope": scope, "previousSpent": previous})

    @staticmethod
    def _sanitize_cost(cost: float) -> float:
        if not isinstance(cost, (int, float)) or isinstance(cost, bool):
            return 0
        if not math.isfinite(float(cost)) or cost < 0:
            return 0
        return float(cost)

    @staticmethod
    def _build_denial(
        reason: EconomicDenialReason,
        cost: float,
        currency: str,
        budget: _BudgetState,
        scope: BudgetScope,
    ) -> EconomicDenialDetails:
        return EconomicDenialDetails(
            reason=reason,
            cost=cost,
            currency=currency,
            budget_scope=scope,
            budget_limit=budget.limit,
            budget_spent=budget.spent,
            budget_remaining=max(0, budget.limit - budget.spent),
            approval_threshold=budget.approval_threshold,
        )

